pallets/referrals/src/lib.rs

```diff
@@ -0,0 +1,674 @@
+		pub fn link_code(origin: OriginFor<T>, code: ReferralCode<T::CodeLength>) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let code = Self::normalize_code(code);
+			let ref_account = Self::referral_account(&code).ok_or(Error::<T>::InvalidCode)?;
+
+			LinkedAccounts::<T>::mutate(who.clone(), |v| -> DispatchResult {
+				ensure!(v.is_none(), Error::<T>::AlreadyLinked);
+
+				ensure!(who != ref_account, Error::<T>::LinkNotAllowed);
+
+				*v = Some(ref_account.clone());
+				Self::deposit_event(Event::CodeLinked {
+					account: who,
+					code,
+					referral_account: ref_account,
+				});
+				Ok(())
+			})?;
+			Ok(())
+		}
+
+		/// Convert accrued asset amount to reward currency.
+		///
+		/// Parameters:
+		/// - `asset_id`: Id of an asset to convert to RewardAsset.
+		///
+		/// Emits `Converted` event when successful.
+		#[pallet::call_index(2)]
+		#[pallet::weight(<T as Config>::WeightInfo::convert())]
+		pub fn convert(origin: OriginFor<T>, asset_id: T::AssetId) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			let asset_balance = T::Currency::balance(asset_id, &Self::pot_account_id());
+			ensure!(asset_balance > 0, Error::<T>::ZeroAmount);
+
+			let total_reward_asset =
+				T::Convert::convert(Self::pot_account_id(), asset_id, T::RewardAsset::get(), asset_balance)?;
+
+			Assets::<T>::remove(asset_id);
+
+			Self::deposit_event(Event::Converted {
+				from: AssetAmount::new(asset_id, asset_balance),
+				to: AssetAmount::new(T::RewardAsset::get(), total_reward_asset),
+			});
+
+			Ok(())
+		}
+
+		/// Claim accumulated rewards
+		///
+		/// IF there is any asset in the reward pot, all is converted to RewardCurrency first.
+		///
+		/// Reward amount is calculated based on the shares of the signer account.
+		///
+		/// if the signer account is referrer account, total accumulated rewards is updated as well as referrer level if reached.
+		///
+		/// Emits `Claimed` event when successful.
+		#[pallet::call_index(3)]
+		#[pallet::weight( {
+			let c = Assets::<T>::iter().count() as u64;
+			let convert_weight = (<T as Config>::WeightInfo::convert()).saturating_mul(c);
+			let w  = <T as Config>::WeightInfo::claim_rewards();
+			let one_read = T::DbWeight::get().reads(1_u64);
+			w.saturating_add(convert_weight).saturating_add(one_read)
+		})]
+		pub fn claim_rewards(origin: OriginFor<T>) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			for (asset_id, _) in Assets::<T>::drain() {
+				let asset_balance = T::Currency::balance(asset_id, &Self::pot_account_id());
+				let r = T::Convert::convert(Self::pot_account_id(), asset_id, T::RewardAsset::get(), asset_balance);
+				if let Err(error) = r {
+					if error == Error::<T>::ConversionMinTradingAmountNotReached.into()
+						|| error == Error::<T>::ConversionZeroAmountReceived.into()
+					{
+						// We allow these errors to continue claiming as the current amount of asset that needed to be converted
```

Comment:
actually, there is a little mistake in here,as i cannot drain all assets anymore. in case of error that we ignore, we need to keep it in the storage.  I need to rework this sligtly.

Reply:
i though that next trade of tat asset will just add it back to the list, so you have chance it will have sufficient amount that time

no need to retry a conversion that failed before right?

Reply:
it will retry when the asset is back inthe list. now it is all good